mm2src/mm2_io/src/fs.rs

```diff
@@ -276,10 +271,51 @@ where
     read_files_with_extension(dir_path, "json").await
 }
 
+/// Creates all the directories along the path to a file.
+pub fn create_parents(path: &dyn AsRef<Path>) -> IoResult<()> {
+    let parent_dir = path.as_ref().parent();
+    let Some(parent_dir) = parent_dir else {
+        return MmError::err(
+            io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!("{} has no parent directory", path.as_ref().display()),
+            ))
+    };
+    if parent_dir.exists() {
+        if !parent_dir.is_dir() {
+            return MmError::err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!("{} is not a directory", parent_dir.display()),
+            ));
+        }
+    } else {
+        fs::create_dir_all(parent_dir)?;
+    }
+    Ok(())
+}
+
+pub fn write(path: &dyn AsRef<Path>, content: &[u8], use_tmp_file: bool) -> IoResult<()> {
+    // Create all the directories in the path.
+    create_parents(path)?;
+    let path_tmp = if use_tmp_file {
+        PathBuf::from(format!("{}.tmp", path.as_ref().display()))
+    } else {
+        path.as_ref().to_path_buf()
+    };
+    // Write the file content into the temp file and then rename the temp file into the desired name.
+    fs::write(&path_tmp, content)?;
+    if use_tmp_file {
+        fs::rename(&path_tmp, path.as_ref())?;
+    }
+    Ok(())
+}
+
 pub async fn write_json<T>(t: &T, path: &Path, use_tmp_file: bool) -> FsJsonResult<()>
 where
     T: Serialize,
 {
+    // FIXME: Create an async counterpart for create_parents? Should we?
+    create_parents(&path).map_err(|err| FsJsonError::IoWriting(err.into_inner()))?;
```

Comment:
Why not [create_dir_all](https://doc.rust-lang.org/beta/std/fs/fn.create_dir_all.html)?

Reply:
I think the names of these functions and comments lack clarity. For example, why aren't we just using `create_dir_all` with `mm_err`? And why does the `write` function behave this way? It's unclear what they do until we look at the implementation details and even then, we don't understand the reasons (of all these conditional behaviours) without checking the other callers...

Reply:
more comments for write() & create_parents()
[bb47ee2](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2398/commits/bb47ee265aae02538e1469c41848da3faa21a8af)

i tried to add more inline comments for `create_parents()` but felt they are just spamy duplicates of the error messages.